src/solution.jl

```diff
@@ -16,113 +13,74 @@ scalar_fn(x::SparseTape) = scalar_fn(to_vaf(x))
-    id_to_variables = context.id_to_variables
-
-    if primal_status != MOI.NO_SOLUTION
-        for (id, var_indices) in var_to_indices
-            var = id_to_variables[id]
-            vexity(var) == ConstVexity() && continue
-            if var_indices isa Tuple
-                vectorized_value_re =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices[1])
-                vectorized_value_im =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices[2])
-                set_value!(
-                    var,
-                    unpackvec(vectorized_value_re, size(var), false) +
-                    im * unpackvec(vectorized_value_im, size(var), false),
-                )
-            else
-                vectorized_value =
-                    MOI.get(model, MOI.VariablePrimal(), var_indices)
-                set_value!(
-                    var,
-                    unpackvec(
-                        vectorized_value,
-                        size(var),
-                        iscomplex(sign(var)),
-                    ),
-                )
-            end
-        end
-    else
-        for (id, var_indices) in var_to_indices
-            var = id_to_variables[id]
-            vexity(var) == ConstVexity() && continue
+    primal_status = MOI.get(context.model, MOI.PrimalStatus())
+    for (id, indices) in context.var_id_to_moi_indices
+        var = context.id_to_variables[id]
+        if vexity(var) == ConstVexity()
+            continue  # Fixed variable
+        elseif primal_status == MOI.NO_SOLUTION
             set_value!(var, nothing)
+        elseif indices isa Tuple  # Complex-valued variable
+            primal_re = MOI.get(p.model, MOI.VariablePrimal(), indices[1])
+            primal_im = MOI.get(p.model, MOI.VariablePrimal(), indices[2])
+            set_value!(
+                var,
+                _unpack_vector(primal_re, size(var)) +
+                im * _unpack_vector(primal_im, size(var)),
+            )
+        else
+            @assert !iscomplex(sign(var))
+            primal = MOI.get(p.model, MOI.VariablePrimal(), indices)
+            set_value!(var, _unpack_vector(primal, size(var)))
         end
     end
-
-    if dual_status != MOI.NO_SOLUTION
-        for (constr, MOI_constr_indices) in pairs(context.constr_to_moi_inds)
-            populate_dual!(model, constr, MOI_constr_indices)
-        end
-    else
-        # Empty duals
-        for constr in keys(context.constr_to_moi_inds)
-            constr.dual = nothing
+    dual_status = MOI.get(context.model, MOI.DualStatus())
+    for (c, indices) in context.constr_to_moi_inds
+        if dual_status == MOI.NO_SOLUTION
+            c.dual = nothing
+        else
+            populate_dual!(context.model, c, indices)
         end
     end
-
-    return nothing
+    return
 end
 
-function populate_dual!(model::MOI.ModelLike, constr::Constraint, ::Nothing)
-    constr.dual = nothing
-    return nothing
```

Comment:
Ditto here: for this method to be called, the element in `.constr_to_moi_inds` would need to be `nothing`, but this can't happen with the current code base. Is it needed for third-parties?

Reply:
I think this is from when there were duals we didn’t populate. Maybe still true for some constraints?

Reply:
Ahhhhh. Okay. That's cause I'm still missing a few tests for `src/constraints/RelativeEntropyEpiCone` and friends. And the current tests only add these indirectly via `QuantumEntropyAtom` etc, so there are no tests for populating duals.

Thanks for the insight. I'm slowly learning the code.

Reply:
The good news is that we're up to 98% coverage :smile: